source/tools/glkernel-cli/KernelJsonImporter.cpp

```diff
@@ -0,0 +1,183 @@
+        }
+
+        for (const auto& widthVariant : *heightArray)
+        {
+            auto widthArray = widthVariant.asArray();
+
+            if (!widthArray)
+            {
+                cppassist::error() << "Malformed kernel input.";
+            }
+
+            for (const auto& elementVariant : *widthArray)
+            {
+                lambda(elementVariant);
+            }
+        }
+    }
+}
+
+// TODO: this is duplicated in the generated code, remove this duplication
+float variantToFloat(const cppexpose::Variant & v)
+{
+    return v.value<float>();
+}
+
+glm::vec2 variantToVec2(const cppexpose::Variant & v)
+{
+    const auto arr = v.asArray();
+
+    const auto x = variantToFloat(arr->at(0));
+    const auto y = variantToFloat(arr->at(1));
+
+    return glm::vec2(x,y);
+}
+
+glm::vec3 variantToVec3(const cppexpose::Variant & v)
+{
+    const auto arr = v.asArray();
+
+    const auto x = variantToFloat(arr->at(0));
+    const auto y = variantToFloat(arr->at(1));
+    const auto z = variantToFloat(arr->at(2));
+
+    return glm::vec3(x,y,z);
+}
+
+glm::vec4 variantToVec4(const cppexpose::Variant & v)
+{
+    const auto arr = v.asArray();
+
+    const auto x = variantToFloat(arr->at(0));
+    const auto y = variantToFloat(arr->at(1));
+    const auto z = variantToFloat(arr->at(2));
+    const auto w = variantToFloat(arr->at(3));
+
+    return glm::vec4(x,y,z,w);
+}
+
+KernelJsonImporter::KernelJsonImporter(const std::string& inputFileName)
+{
+    cppexpose::Variant root;
+
+    if (!cppexpose::JSON::load(root, inputFileName))
+    {
+        cppassist::error() << "Input JSON file could not be loaded.";
+    }
+
+    auto rootMap = root.asMap();
+    if (!rootMap)
+    {
+        cppassist::error() << "Input JSON is malformed.";
+    }
+
+    auto depthArray = rootMap->at("kernel").asArray();
+    auto sizeMap = rootMap->at("size").asMap();
+
+    int depth = sizeMap->at("depth").value<int>();
+    int height = sizeMap->at("height").value<int>();
+    int width = sizeMap->at("width").value<int>();
+    size_t numCells = 0;
+    int numComponents = -1;
+
+    forEachCell(depthArray, [&numCells, &numComponents](const cppexpose::Variant& elementVariant) {
+        if (elementVariant.isFloatingPoint())
+        {
+            if (numComponents != -1 && numComponents != 1) {
+                cppassist::error() << "All cells must have the same cell type (float, vec2, vec3 or vec4).";
+            }
+
+            numComponents = 1;
+        }
+        else
+        {
+            if (!elementVariant.isArray()) {
+                cppassist::error() << "Cell is not floating point or array.";
+            }
+
+            if (numComponents != -1 && elementVariant.asArray()->size() != numComponents) {
+                cppassist::error() << "All cells must have the same cell type (float, vec2, vec3 or vec4).";
+            }
+
+            numComponents = elementVariant.asArray()->size();
+        }
+
+        numCells++;
```

Comment:
numCells is never used, since you do forEachCell instead of a `for (i = 0; i < numCells)` kind of thing